### PR TITLE
Optimize min and max(reduce_dim) performance on CPU

### PR DESCRIPTION
**Pull request by XiaobingSuper**
This PR is about improve min and max(reduce_dim) performance on CPU.  
Test script:
```
 import torch
 import torch.nn as nn
 import time

 torch.manual_seed(0)

 def _time():
     return time.time()

 device = "cpu"
 torch.set_num_threads(1)
 
 #warm up
 for n in [10, 200]:
     # contiguous
     # input = torch.randn(n, n, n, requires_grad=False, device=device)
     # discontiguous
     input = torch.randn(n, 2*n, n, requires_grad=False, device=device)[:, :n, :]
     for dim in range(input.dim()):
         for i in range(1000):
             output = input.min(dim)
             #output = input.max(dim)

 for n in [10, 200]:
     # contiguous input
     # input = torch.randn(n, n, n, requires_grad=False, device=device)
     # discontiguous
     input = torch.randn(n, 2*n, n, requires_grad=False, device=device)[:, :n, :]
     for dim in range(input.dim()):
         fwd_t = 0
         for i in range(10000):
             t1 = _time()
             output = input.min(dim)
             #output = input.max(dim)
             t2 = _time()
             fwd_t = fwd_t + (t2 -t1)
         fwd_avg = fwd_t / 10000 * 1000
         print("size = (%d, %d, %d); reduce dim=%d; compute time is %.4f(ms)" % (n, n, n, dim, fwd_avg))
```
Test device: **skx-8180**.

### Contiguous case.
- num_threads = 56

  | Bef(ms) | Bef(ms) | Bef(ms) | Aft(ms) | Aft(ms) | Aft(ms)
-- | -- | -- | -- | -- | -- | --
size | dim=0 | dim=1 | dim=2 | dim=0 | dim=1 | dim=2
n=10 | 0.0243 | 0.0243 | 0.0244 | 0.0063 | 0.0065 | 0.0063
n=200 | 0.9615 | 0.9453 | 0.7772 | 0.2937 | 0.2675 | 0.2607

- num_threads = 1

  | Bef(ms) | Bef(ms) | Bef(ms) | Aft(ms) | Aft(ms) | Aft(ms)
-- | -- | -- | -- | -- | -- | --
size | dim=0 | dim=1 | dim=2 | dim=0 | dim=1 | dim=2
n=10 | 0.0126 | 0.0126 | 0.0114 | 0.0062 | 0.0065 | 0.0064
n=200 | 32.1276 | 33.3489 | 29.0757 | 8.0556 | 7.0188 | 6.5014

### Discontiguous case.
- num_threads = 56

  | Bef(ms) | Bef(ms) | Bef(ms) | Aft(ms) | Aft(ms) | Aft(ms)
-- | -- | -- | -- | -- | -- | --
size | dim=0 | dim=1 | dim=2 | dim=0 | dim=1 | dim=2
n=10 | 0.0106 | 0.0115 | 0.0131 | 0.0063 | 0.0066 | 0.0065
n=200 | 14.652 | 15.3496 | 9.8153 | 0.2946 | 0.2708 | 0.267
- num_threads = 1

  | Bef(ms) | Bef(ms) | Bef(ms) | Aft(ms) | Aft(ms) | Aft(ms)
-- | -- | -- | -- | -- | -- | --
size | dim=0 | dim=1 | dim=2 | dim=0 | dim=1 | dim=2
n=10 | 0.0108 | 0.0116 | 0.0132 | 0.0058 | 0.0062 | 0.0061
n=200 | 12.5132 | 13.0785 | 9.6738 | 8.3733 | 7.3051 | 6.4566

https://github.com/pytorch/pytorch/issues/24671 and https://github.com/pytorch/pytorch/issues/24672 are also fixed.

### CODE REVIEW

**Comment by ngimel** [comment]
The difference between perf numbers for the different dimensions becomes much more pronounced as tensor becomes too big to fit in cache, e.g. for 256 here's what I'm getting for a single thread:
```
size = (256, 256, 256); reduce dim=0; compute time is 54.9099(ms)
size = (256, 256, 256); reduce dim=1; compute time is 38.0527(ms)
size = (256, 256, 256); reduce dim=2; compute time is 17.4140(ms)
```
for comparison, vectorized reduction used for sum handles different dimensions better (using the same script, replacing `max` with `sum`), both in absolute and relative terms
```
size = (256, 256, 256); reduce dim=0; compute time is 8.6444(ms)
size = (256, 256, 256); reduce dim=1; compute time is 8.1093(ms)
size = (256, 256, 256); reduce dim=2; compute time is 5.4094(ms)
```
Still, this PR provides great improvement compared to current implementation, so we can go ahead with it.

**Comment by ngimel** [comment]
I suggest factoring all the logic out of min_kernel_impl and max_kernel_impl, like here (this commit is on top of your PR) https://github.com/ngimel/pytorch/commit/734c094a8cd9bfd1f9758d1a21b0f3f776ee409d, so that all the logic is contained in compare_base_kernel
I've checked that single-threaded benchmarking numbers are the same, but did not do any more testing.

**Comment by XiaobingSuper** [comment]
@ngimel, I test your patch ngimel@734c094, I find there has a big performance degradation for single-threaded performance in my side, see(test min case with contiguous input):
For my change:
```
size = (10, 10, 10); reduce dim=0; compute time is 0.0061(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0063(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0062(ms)
size = (200, 200, 200); reduce dim=0; compute time is 7.9776(ms)
size = (200, 200, 200); reduce dim=1; compute time is 7.0235(ms)
size = (200, 200, 200); reduce dim=2; compute time is cd (ms)
```
Apply your patch:
```
size = (10, 10, 10); reduce dim=0; compute time is 0.0062(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0065(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0064(ms)
size = (200, 200, 200); reduce dim=0; compute time is 10.1095(ms)
size = (200, 200, 200); reduce dim=1; compute time is 10.0302(ms)
size = (200, 200, 200); reduce dim=2; compute time is 9.4876(ms)
```
So I decide to not change the code now. Thanks!

**Review by ngimel — APPROVED** [review]

**Review by VitalyFedyunin — APPROVED** [review]
